apps/emqx/src/emqx_trace/emqx_trace.erl

```diff
@@ -83,6 +87,27 @@ unsubscribe(<<"$SYS/", _/binary>>, _SubOpts) ->
 unsubscribe(Topic, SubOpts) ->
     ?TRACE("UNSUBSCRIBE", "unsubscribe", #{topic => Topic, sub_opts => SubOpts}).
 
+rendered_action_template(ActionID, RenderResult) ->
+    Msg = lists:flatten(io_lib:format("action_template_rendered(~ts)", [ActionID])),
+    TraceResult = ?TRACE("QUERY_RENDER", Msg, RenderResult),
+    case logger:get_process_metadata() of
+        #{stop_action_after_render := true} ->
+            %% We throw an unrecoverable error to stop action before the
+            %% resource is called/modified
+            StopMsg = lists:flatten(
+                io_lib:format(
+                    "action_stopped_after_render(~ts): "
+                    "Action stopped after template render due to test setting.",
+                    [ActionID]
+                )
+            ),
+            MsgBin = iolist_to_binary(StopMsg),
```

Comment:
nit: for "~ts", it is not guaranteed to be iolist. must call unicode:characters_to_binary instead.
however, given that ActionID is restricted string, it's ok.

last time someone had to find all the ~s in the code base, change to ~ts, if we take care of this in general, should save some time when this has to be done again.